### ISSUE TITLE
Testing helpers

### ISSUE THREAD

**Issue by Whoaa512** [jondot/hygen#157]
Thanks for building this awesome code gen tool, and great job on v5.

What's the current strategy for unit testing generators?

Is there some way to load run a generator with a mocked filesystem? Perhaps even being able to seed that file system with dummy data that the generator would interact with.

I'm planning to introduce hygen at my company as a yeoman replacement, but testing is a desired feature.

I'm sure something could be hacked together, but I was curious if you've had any thoughts or ideas already.

**Comment by jondot** [comment]
Interesting, actually hygen has `metaverse` (in `__tests__`) which is hygen's test "universe". Hygen tests itself using a similar technique to what you need, on the exception of using mock filesystems, that were removed because they created a very complex maintenance story.

Having that in mind, hygen departs from Yeoman in the testing approach too. Which is to say: don't test the generator, test the generator output.

So here's a make-shift testing approach:

```
$ hygen component new --name foobar && yarn test foobar
```
We can build some privisions to tell hygen to:

* Generate in "test" mode, which means put material in `/tmp/random-id`
* Pick up input parameters from a file (and by that have a _fixture_ like concept)

Another, different approach would be to package Hygen's test harness (metaverse) into a tool which means what you'll get is the following:

* Pick up input params and command from a file
* Generate
* Compare output to a well-known "gold-set" of what you expect output to be
* Do all this work in `/tmp`

Any of these make sense to you? if so how would you see the optimal solution?

**Comment by Whoaa512** [comment]
In my codebase, I actually ended doing something similar to what `metaverse` does.

Basically I created a mock version of what the `bin.ts` files does and just have before/after hooks to setup/clean the tmp folder I'm using. 

https://gist.github.com/Whoaa512/f65902089923ba4c9eb4caf1c9560557

I was manually comparing folder contents and using jest.snapshots, but I like your pattern of using `dir-compare` and just having the expected output in git.
